### PR TITLE
feat: Add HuggingFace runtime out-of-the-box support

### PR DESCRIPTION
**Pull request by terrytangyuan**
Part of https://github.com/kserve/kserve/issues/3393.

- [x] Tests
- [x] Update Python spec
- [x] Regenerate manifests

### CODE REVIEW

File: config/runtimes/kserve-huggingfaceserver.yaml [hunk 1]
```diff
@@ -0,0 +1,28 @@
+apiVersion: serving.kserve.io/v1alpha1
```
Comment (sivanantha321) [inline config/runtimes/kserve-huggingfaceserver.yaml:1]: This also needs to be added in helm charts. https://github.com/kserve/kserve/blob/master/charts/kserve-resources/templates/clusterservingruntimes.yaml

File: pkg/apis/serving/v1beta1/inference_service_defaults.go [hunk 2]
```diff
@@ -219,6 +222,20 @@ func (isvc *InferenceService) assignONNXRuntime() {
...
 
+func (isvc *InferenceService) assignHuggingFaceRuntime() {
+	// assign protocol version 'v2' if not provided for backward compatibility
+	if isvc.Spec.Predictor.HuggingFace.ProtocolVersion == nil {
+		protocolV2 := constants.ProtocolV2
+		isvc.Spec.Predictor.HuggingFace.ProtocolVersion = &protocolV2
+	}
+	isvc.Spec.Predictor.Model = &ModelSpec{
+		ModelFormat:            ModelFormat{Name: constants.SupportedModelHuggingFace},
+		PredictorExtensionSpec: isvc.Spec.Predictor.ONNX.PredictorExtensionSpec,
```
Comment (sivanantha321) [inline pkg/apis/serving/v1beta1/inference_service_defaults.go]: Why are we using ONNX ?

Comment (terrytangyuan) [reply pkg/apis/serving/v1beta1/inference_service_defaults.go]: @sivanantha321 Thanks. This is still WIP and haven’t tested yet. Please do not review yet.

File: pkg/apis/serving/v1beta1/predictor.go [hunk 3]
```diff
@@ -44,6 +44,8 @@ type PredictorSpec struct {
 	Triton *TritonSpec `json:"triton,omitempty"`
 	// Spec for ONNX runtime (https://github.com/microsoft/onnxruntime)
 	ONNX *ONNXRuntimeSpec `json:"onnx,omitempty"`
+	// Spec for HuggingFace runtime (https://github.com/huggingface)
+	HuggingFace *HuggingFaceRuntimeSpec `json:"huggingface,omitempty"`
```
Comment (yuzisun) [inline pkg/apis/serving/v1beta1/predictor.go:48]: Actually we are trying to deprecate these individually defined runtime types as we are moving users to the model spec instead. @sivanantha321 has a PR for this https://github.com/kserve/kserve/pull/3144

Comment (terrytangyuan) [reply pkg/apis/serving/v1beta1/predictor.go:48]: Yeah I am aware of that PR and thought about it as well. IMO we should include this for consistency (better UX) and then deprecate all the individually defined runtimes together later as part of https://github.com/kserve/kserve/pull/3144.

File: charts/kserve-resources/templates/clusterservingruntimes.yaml [hunk 4]
```diff
@@ -344,3 +344,32 @@ spec:
         limits:
           cpu: "1"
           memory: 2Gi
+---
+apiVersion: serving.kserve.io/v1alpha1
+kind: ClusterServingRuntime
+metadata:
+  name: kserve-huggingfaceserver
```
Comment (terrytangyuan) [inline charts/kserve-resources/templates/clusterservingruntimes.yaml:351]: Note to myself: change this to `kserve-huggingface`

**Comment by yuzisun** [comment]
@terrytangyuan Looks great! can you help add an e2e test?

**Pull request merged**